### ISSUE TITLE
Could it support generic patching?

### ISSUE THREAD

**Issue by HMaker** [Ceiridge/Chrome-Developer-Mode-Extension-Warning-Patcher#81]
I want to automate multiple Chrome windows by sending window messages for mouse and keyboard input (WM_MOUSEMOVE, WM_LBUTTONDOWN, WM_KEYDOWN, WM_MOUSEWHEEL, ...), but Chrome windows on background ignores WM_MOUSEWHEEL messages. that's a problem because I can't have mutiple windows on foreground...

I found the code responsible for that at
https://source.chromium.org/chromium/chromium/src/+/main:ui/views/win/hwnd_message_handler.cc;l=3179
the function [RerouteMouseWheel][1] returns `true` when the message must be ignored, I need to make it return always `false`.

I know this patcher was made for specific use-case, but I found it takes bytes patterns to find the place to patch, so couldn't this patcher be made to support generic patching where people could provide any `patterns.xml` file?

[1]: https://source.chromium.org/chromium/chromium/src/+/main:ui/base/win/mouse_wheel_util.cc;l=59

**Comment by Ceiridge** [comment]
Yes, it can patch anything in the executable memory of *chrome.dll*. It first searches for the pattern, which should be unique. It consists of the assembled bytes of the function's assembly instructions with other addresses and changing values as wildcards.
Once the pattern is found, a patch is applied at a given offset (can be 0, so it is directly overwritten where the pattern has been found). Normally, only one byte is patched, but you can also enter multiple assembled x86 instruction bytes (for example: 
```asm
xor rax, rax
ret
```
to return false, as seen in the *Remove QR Code Availability* Patch).
If you want to contribute a patch, you can do it like this. I would then also recommend reading the comments in the [patterns.xml](https://github.com/Ceiridge/Chrome-Developer-Mode-Extension-Warning-Patcher/blob/master/patterns.xml) file, although they are annoying to understand

You can also use a local patterns.xml file by starting the patcher installer in debug mode, as far as I remember

**Comment by HMaker** [comment]
@Ceiridge Great. So what means `orig`, `patch`, `sig` and `sigOffset` attributes of `PatchData` tag?
```xml
<Patch pattern="Remove QR Code Availability" group="5">
	<PatchData type="x64" orig="0xE9" patch="0x00" sig="true" sigOffset="0x0">
		<Offset>0x22</Offset> <!-- Offset to get to jmp + 1 -->
		<NewBytes>4831C0C3</NewBytes> <!-- xor rax, rax; ret; (= return 0;) -->
	</PatchData>
</Patch>
```

**Comment by Ceiridge** [comment]
**Type**: x64 (x86-64). The patcher used to also support x86-32, so don't use x86
**Orig**: Original byte that should be replaced. 0xFF is a wildcard, so it would be ignored then. If the byte does not match, an error is thrown
**Patch**: The new byte
**Sig**: Is the pattern a *signature*? This is rather complicated to explain and it's very rarely used. It's sometimes more reliable to find an address of a function by searching for an instruction that has an address to it (for example call FunctionAddress). Then, if sig is enabled, it takes that address offset of the call instuction for example and calculates the right address.
**SigOffset**: If sig is enabled, this exact offset will be added to the followed address as explained above

**Offset**: This is added to the found pattern location, can also be 0x0
**NewBytes**: Optional, but this allows you to write multiple instructions easily. In this case: `0x48 31 C0` = `xor rax, rax` and `0xC3` = `ret`)

**Comment by Ceiridge** [comment]
If you want to find out more about how signatures work (which I do not recommend) or something else in general: https://github.com/Ceiridge/Chrome-Developer-Mode-Extension-Warning-Patcher/blob/master/ChromePatcherDll/patches.cpp#L231
